compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/AggregatedOpInterfaceImpl.cpp

```diff
@@ -225,4 +261,265 @@ OnlineAttentionOp::decomposeOperation(OpBuilder &b) {
+  if (constKTileSize) {
+    kTileSize = b.getIndexAttr(constKTileSize.value());
+  }
+  SmallVector<OpFoldResult> inputSizes =
+      tensor::getMixedSizes(b, loc, getInput());
+  // Find the innermost non-batch dimension. This dimension is the fastest
+  // changing dimension with the K dimension of the im2col iteration domain.
+  // This means it is the innermost dimension of the extract_slice on the
+  // input tensor, and the slice wants to be contiguous along this dimension.
+  SetVector<int64_t> batchPosSet(getBatchPos().begin(), getBatchPos().end());
+  OpFoldResult innerSliceSize;
+  for (int idx = inputSizes.size() - 1; idx >= 0; --idx) {
+    if (!batchPosSet.contains(idx)) {
+      innerSliceSize = inputSizes[idx];
+      break;
+    }
+  }
+  bool tileK =
+      !willBeContiguousSlice(innerSliceSize, kTileSize, kOffset.front());
+  if (!tileK) {
+    iterationDomain.pop_back();
+  } else {
+    kTileSize = b.getIndexAttr(1);
+  }
+
+  // Build loop nest.
+  SmallVector<Value> lbs, ubs, steps;
+  for (auto range : iterationDomain) {
+    lbs.push_back(getValueOrCreateConstantIndexOp(b, loc, range.offset));
+    ubs.push_back(getValueOrCreateConstantIndexOp(b, loc, range.size));
+    steps.push_back(getValueOrCreateConstantIndexOp(b, loc, range.stride));
+  }
+  scf::LoopNest loopNest = scf::buildLoopNest(
+      b, loc, lbs, ubs, steps, getOutput(),
+      [&](OpBuilder &nestedBuilder, Location loc, ValueRange outputIvs,
+          ValueRange iterArgs) -> scf::ValueVector { return iterArgs; });
+  SmallVector<Value> ivs;
+  for (scf::ForOp loop : loopNest.loops) {
+    ivs.push_back(loop.getInductionVar());
+  }
+
+  // Step 2: Compute indices into the input tensor for extract_slice.
+  b.setInsertionPoint(loopNest.loops.front());
+  SetVector<int64_t> mPosSet(getMPos().begin(), getMPos().end());
+
+  // Compute the basis for the iteration space of the convolution window
+  // (i.e., the H and W dims of the convolution output).
+  SmallVector<Value> mBasis;
+  ArrayRef<int64_t> strides = getStrides();
+  ArrayRef<int64_t> dilations = getDilations();
+  SmallVector<OpFoldResult> kernelSize = getMixedKernelSize();
+  for (auto [idx, pos] : llvm::enumerate(getMPos())) {
+    AffineExpr x, k;
+    bindDims(getContext(), x, k);
+    AffineExpr mapExpr =
+        (x - 1 - (k - 1) * dilations[idx]).floorDiv(strides[idx]) + 1;
+    OpFoldResult size = affine::makeComposedFoldedAffineApply(
+        b, loc, AffineMap::get(2, 0, {mapExpr}, getContext()),
+        {inputSizes[pos], kernelSize[idx]});
+    mBasis.push_back(getValueOrCreateConstantIndexOp(b, loc, size));
+  }
+
+  // Delinearize the k_offset into an offset into the convolution window and
+  // any reduced channels. For an NHWC conv2d, the basis for delinearization
+  // would be [P, Q, C] for a PxQ kernel with C channels.
+  Location nestedLoc =
+      loopNest.loops.back().getBody()->getTerminator()->getLoc();
+  b.setInsertionPointToStart(loopNest.loops.back().getBody());
```

Comment:
I think we need to add a `OpBuilder::InsertionGuard guard(b)`. Life is easier if the callee does not mess up the insertion point. Otherwise, callers will always need to look at impl details or always create a guard, which looks bad to me.